[Reproducers] Add FileCollector support to DependencyTracker

JDevlieghere opened the pull request:
For reproducers in LLDB, we need to collect module dependency
information coming from the clang importer. However, we cannot override
the dependency collector, like we do in LLDB, because its interface is
completely hidden in the clang importer. The solution is to pass the
FileCollector through the DependencyTracker.

JDevlieghere committed:
[Reproducers] Add FileCollector support to DependencyTracker

For reproducers in LLDB, we need to collect module dependency
information coming from the clang importer. However, we cannot override
the dependency collector, like we do in LLDB, because its interface is
completely hidden in the clang importer. The solution is to pass the
FileCollector through the DependencyTracker.

JDevlieghere commented:
LLDB PR: https://github.com/apple/swift-lldb/pull/1836

jrose-apple commented:
I'm confused here. So now we're collecting the same list of files twice? That seems like there's a better answer somewhere else.

Also, this should probably go to `master` rather than `master-next`, no?

JDevlieghere commented:
> I'm confused here. So now we're collecting the same list of files twice? That seems like there's a better answer somewhere else.

Yes and no. The FileCollector for LLDB is global and we create more than once instance of the ClangImporter. The list of files in the FileCollector is the union of all the files seen by the clang importer (and everything else in LLDB). 

I'm haven't checked this, but I can also imagine that the files needed by the reproducers don't have to be identical to the Swift dependency files. For example, in order to guarantee that clang behaves the same way during replay, we might have to capture module maps that aren't used. 

Finally, I'm sure we could switch over Swift to use the FileCollector as well, though it would be more heavy-weight for the regular (non-reproducer) scenario. Currently a list of files is all we need (because we only generate the dependency files from it?) while the FileCollector keeps track of file mappings, to be able to write a VFS overlay in the end. 

However, even if we did that, we'd need to _pull in_ that information from Swift somehow. With the ClangImporter using LLDB's file collector, we're essentially _being pushed_ the information as soon as it's available, similar to how clang does this. That makes things a lot easier when we're in a signal handler after a crash.  

> Also, this should probably go to `master` rather than `master-next`, no?

I can't work on master because I have dependencies in clang and llvm that shouldn't end up on the release branch. Technically stable could work, but that'd mean they'd no longer be aliases for the release branch and at this point in the release cycle it seems easier to just work on master-next.

jrose-apple commented:
Yeah, okay. LLVM and Clang `stable` really do benefit from staying basically the same as the release branch.

Can I take a look at the solution in Clang? If LLDB's using a custom subclass of DependencyCollector there, maybe Swift should either conditionally or always do that too. Or it could chain to one of those rather than talking directly to FileCollector. The way you've set things up here is *okay,* but it does feel ad hoc. (And doesn't have any tests in the Swift repo.)

JDevlieghere commented:
> Can I take a look at the solution in Clang? If LLDB's using a custom subclass of DependencyCollector there, maybe Swift should either conditionally or always do that too. Or it could chain to one of those rather than talking directly to FileCollector. The way you've set things up here is _okay,_ but it does feel ad hoc. (And doesn't have any tests in the Swift repo.)

Yeah, for Clang we just subclass the `ModuleDependencyCollectorAdaptor` and basically make everything a NOOP, except for adding the files to our `FileCollector`. I'd be totally fine doing that here too, but it seems like we made quite the effort to hide the `ClangImporterDependencyCollector`, presumably because you don't want to leak the corresponding Clang header from `ModuleLoader.h `.

jrose-apple commented:
Yeah, that's pretty much the reason. (Also, hm, why ModuleDependencyCollector specifically for yours?)

Okay, I guess you're right: it makes more sense for Swift to expose an API that uses llvm::FileCollector than it does to expose one that uses clang::DependencyCollector, and it's not really that much extra work to make the other list we're not going to use.

harlanhaskins commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -319,11 +320,13 @@ class BridgingPPTracker : public clang::PPCallbacks {
 class ClangImporterDependencyCollector : public clang::DependencyCollector
 {
   llvm::StringSet<> ExcludedPaths;
+  std::shared_ptr<llvm::FileCollector> FileCollector;
```
Can you leave a comment here explaining that this is for LLDB to generate reproducers and does not represent a precise view of a module's dependencies?

JDevlieghere committed:
[ClangImporter] Add comment to the ClangImporterDependencyCollector

Add a comment to the ClangImporterDependencyCollector explaining that
the FileCollector is used by LLDB to generate reproducers and is not
involved in dependency collection in Swift.
